### PR TITLE
[jit] Switch mutable lists to new mutable schema

### LINKED ISSUE #12780: [pytorch/jit] all lists get merged during tracing
## 🐛 Bug

If you define a nn.Module that contains a forward func with multiple lists (That you stack together into a tensor at some point) that process works perfectly fine in the conventional workflow. 
However, if you attempt to convert your module into a ScriptModule that will be compiled into a graph representation, every list object defined in a `script_method` function seem to merge together somehow.

## To Reproduce

Create a test bench with the source code below, it should be fully self contained
```python


import torch
from torch import nn
from torch.functional import F

class ACTNN(torch.jit.ScriptModule):

    __constants__ = ['hidden_depth', 'hidden_width']
    # hidden_width = 125
   # hidden_depth = 3
    # output_size = 80
    # input_size = 125
    def __init__(self, input_size, hidden_width, output_size, hidden_depth):
        super(ACTNN, self).__init__()
        self.hidden_depth = hidden_depth
        self.hidden_width = hidden_width
        self.rnn = torch.jit.trace(nn.RNN(input_size, hidden_width,
                                          num_layers=self.hidden_depth, batch_first=True), example_inputs=(torch.rand(1, 1, input_size), torch.rand(self.hidden_depth, 1, hidden_width)))

        self.proc = torch.jit.trace(nn.Linear(hidden_width, output_size), example_inputs=torch.rand(1, 1, hidden_width))


    @torch.jit.script_method
    def forward(self, input: torch.Tensor, hidden: torch.Tensor):

        input = input.view(1, 1, -1)
        hidden = hidden.view(self.hidden_depth, 1, self.hidden_width)
        # Define the list objects
        states = []
        hiddens = []
        halt_probs = []
        n = 0
        rnn_out, hidden = self.rnn(input, hidden)
        state = self.proc(rnn_out)
        hiddens.append(hidden)
        states.append(state)
        halt_probability = F.sigmoid(hiddens[n].sum())
        halt_probs.append(halt_probability)
       # At this point, both the states & halt_probs lists will contain:
       # 1x 1,1,80 wide tensor
       # 1x 3,1,125 wide tensor
       # 1x 1 wide tensor

        print(states)
        print(halt_probs)
        residual = torch.sum(torch.cat(halt_probs))

        states_tensor = torch.stack(states, dim=1)
        hiddens_tensor = torch.stack(hiddens, dim=1)
        halt_subset = halt_probs[0:n-1]
        halt_subset.append(residual)
        halt_prob_tensor = torch.stack(halt_subset).view(-1)
        output = torch.mv(states_tensor, halt_prob_tensor)
        hidden = torch.mv(hiddens_tensor, halt_prob_tensor)
        return output, hidden
```
dimensions of elements in `states`:
```
[ Variable[CPUFloatType]{3,1,125} ],
[ Variable[CPUFloatType]{1,1,80} ],
[ Variable[CPUFloatType]{1} ]]
```

and here is the error message that causes an overall failure of the trace operation:
```
RuntimeError: 
invalid argument 0: Sizes of tensors must match except in dimension 1. Got 3 and 1 in dimension 0 at /pytorch/aten/src/TH/generic/THTensorMoreMath.cpp:1317:
operation failed in interpreter:
    halt_probability = self.halt_noise(halt_probability, training)
    halt_probs.append(halt_probability)
    print(halt_probs)
    residual = torch.sum(torch.cat(halt_probs))
    states_tensor = torch.stack(states, dim=1)
                    ~~~~~~~~~~~ <--- HERE
```

## Expected behavior

for the above code sample, 
`hiddens` should contain:
```
[ Variable[CPUFloatType]{3,1,125} ],
```
`states` should contain:
```
[ Variable[CPUFloatType]{1,1,80} ]
```
`halt_probs` should contain:
```
[ Variable[CPUFloatType]{1} ]]
```
.
## Environment

 - PyTorch Version (e.g., 1.0): 1.0.0.dev20181004
 - OS (e.g., Linux): Ubuntu 16.04
 - How you installed PyTorch (`conda`, `pip`, source): `pip install /tmp/torch_nightly-1.0.0.dev20181004-cp35-cp35m-linux_x86_64.whl`
 - Build command you used (if compiling from source): N/A
 - Python version: 3.5.4
 - CUDA/cuDNN version: N/A
 - GPU models and configuration: N/A
 - Any other relevant information: N/A

## Additional context

As I am doing some somewhat advanced pytorch module development, I could understand if my usage here is not supported - however there is currently no error message saying that I can't construct multiple tensors from elements captured in a forward pass.

It's certainly possible that some other part of my algorithm is failing, or maybe I'm using torch.jit.ScriptModule incorrectly? That would be great if that was the case,

I could see this posing issues down the road with other types of dynamic process flow during forward pass loops (like ACT, but also NTMs, etc), which are probably mechanisms we want pytorch 1.0 to support properly.

### PR DESCRIPTION
**Pull request by suo**
Goodbye, World! This PR removes the world tokens and associated pass and switches lists over to the new mutability/aliasing annotations.

Should resolve #12780 since we are disabling optimization pending alias analysis.

### CODE REVIEW

File: torch/csrc/jit/graph_executor.cpp [hunk 1]
```diff
@@ -331,17 +330,13 @@ struct GraphExecutorImpl {
...
-      // until we have correct alias analysis any use of mutable operators
-      // disables all optimization
-      if(hasMutableOperators(this->graph->block())) {
-        optimize = false;
-      }
-    }
+      : graph(prepareGraph(graph)),
+        // until we have correct alias analysis any use of mutable operators
+        // disables all optimization
+        optimize(!hasMutableOperators(this->graph->block())),
```
Comment (suo) [inline torch/csrc/jit/graph_executor.cpp]: btw, here's an example of where we should have followed the FB style guide; `optimize` was getting shadowed in the constructor and so we weren't actually setting it. Using `optimize_` to denote the member variable status would have disambiguated

Comment (zdevito) [reply torch/csrc/jit/graph_executor.cpp]: good catch! The current code you have here ignores the `optimize` parameter, though.

File: torch/csrc/jit/register_prim_ops.cpp [hunk 2]
```diff
@@ -791,7 +767,7 @@ Operator(                                                                      \
         "aten::slice(" decl_type "[] l, int start, int end=9223372036854775807, int step=1) -> " decl_type "[]", \
         listSlice<Shared<c_type>, c_type::ElemType>), \
     Operator( \
-        "aten::append(World w, " decl_type "[] self, " decl_type " el) -> World", \
+        "aten::append(" decl_type "[](a!) self, " decl_type " el) -> " decl_type "[](a!)", \
```
Comment (zdevito) [inline torch/csrc/jit/register_prim_ops.cpp]: This annotation is pretty tricky and is outside what the annotation language can parse. Example:

```
aten::append(Tensor(b)[](a!) self, Tensor(c) el) -> Tensor(b|c)[](a!)
# but the alias set of self's elements _also_ changes after this instruction to include set c as well, we do
# not have a way to denote this.
```

For now we can leave it as it is written, but we need to be careful that list extracts always say the resulting Tensor can be in any alias set until these operators capture this behavior more accurately.

**Review by zdevito — APPROVED** [review]
Looks good! One small possible bug.

**Pull request closed without merging**